x/gamm/keeper/pool.go

```diff
@@ -79,6 +79,72 @@ func (k Keeper) SetPool(ctx sdk.Context, pool types.PoolI) error {
 	return nil
 }
 
+// CleanupBalancerPool destructs a pool and refund all the assets according to
+// the shares held by the accounts. CleanupBalancerPool should be called not during
+// the chain execution time, as it iterates the entire account balances.
+func (k Keeper) CleanupBalancerPool(ctx sdk.Context, poolId uint64) (err error) {
+	pool, err := k.GetPool(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	totalShares := pool.GetTotalShares().Amount
+	shareDenom := pool.GetTotalShares().Denom
+	poolAssets := pool.GetAllPoolAssets()
+
+	// first iterate through the share holders and burn them
+	k.bankKeeper.IterateAllBalances(ctx, func(addr sdk.AccAddress, coin sdk.Coin) (stop bool) {
+		if coin.Denom != shareDenom || coin.Amount.IsZero() {
+			return
+		}
+
+		shareAmount := coin.Amount
+
+		// Burn the share tokens
+		err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, addr, types.ModuleName, sdk.Coins{coin})
+		if err != nil {
+			return true
+		}
+
+		err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.Coins{coin})
+		if err != nil {
+			return true
+		}
+
+		// Refund assets
+		pool.SubTotalShares(shareAmount)
+		for _, asset := range poolAssets {
+			err = k.bankKeeper.SendCoins(
+				ctx, pool.GetAddress(), addr, sdk.Coins{{asset.Token.Denom, asset.Token.Amount.Mul(shareAmount).Quo(totalShares)}})
+			if err != nil {
+				return true
+			}
+		}
+
+		return false
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// sanity check
+	if !pool.GetTotalShares().IsZero() {
+		panic("pool total share should be zero after cleanup")
+	}
```

Comment:
Can zero share be assured everytime(division factor etc.)?
My only concern is that it might not: considering the fact that I remember that somewhere in the code what we did was iterate through coins and send a part of it every iteration, and at the last iteration, we would send the all the remainings.

Reply:
Great point, I originally thought the total shares are only added/subtracted, not divided; I will check if there are any case of division.

Reply:
Is this fixed? / checked?

Reply:
Because shares are just a sdk.Coin, it could be IBC'd out or sent to module logics, and stay under the module account. I am pretty sure if we count all the address held shares and module held shares, it would sum up to GetTotalShares(). Let me add the logic for module accounts to(or, is it already iterated by GetAllBalances?)

Reply:
Can you investigate if module accounts are iterated over? I think we should definitely exclude certain module accounts, e.g. IBC

Reply:
The iteration do include the module accounts. I will exclude the module account held LP tokens.

Reply:
Okay, annoyingly, there is no way to distinguish module address and normal address by only using the address itself. We could read all the module address and compare to them each time iterating over the list of address, but at this stage it feels little adhoc. Let me think about some alternatives a bit.